flang/lib/Lower/ConvertExprToHLFIR.cpp

```diff
@@ -1435,8 +1434,12 @@ struct UnaryOp<
       // allocate space on the stack for toBuffer
       auto dest = builder.create<fir::AllocaOp>(loc, toTy,
                                                 mlir::ValueRange{bufferSize});
-      builder.create<fir::CharConvertOp>(loc, lhs.getFirBase(), origBufferSize,
-                                         dest);
+      auto src = hlfir::convertToAddress(loc, builder, lhs,
+                                         lhs.getFortranElementType());
+      builder.create<fir::CharConvertOp>(loc, src.first.getCharBox()->getAddr(),
+                                         origBufferSize, dest);
+      if (src.second.has_value())
+        src.second.value();
```

Comment:
Are you sure this is calling the cleanup? I would actually expect `src.second.value()()` or `(*src.second)()` here.

Can you ensure there is an hlfir.end_associate created after the fir.char_convert for the following example:

```
subroutine charconvert3(c, c4)
  character(kind=1, len=*) :: c
  character(kind=4, len=*) :: c4
  c4 = c // c
end subroutine
```

Reply:
Yep, you are correct. I didn't actually call the `hlfir::cleanupFunction`.  I added another set of parens so now it's 
```cpp
src.second.value()()
```

I added your concat test and verified that `hlfir.end_associate` is called